azure-core/azure/core/pipeline/policies/custom_hook.py

```diff
@@ -0,0 +1,46 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+"""
+This module is the requests implementation of Pipeline ABC
+"""
+from .base import SansIOHTTPPolicy
+
+class CustomHookPolicy(SansIOHTTPPolicy):
+    """A simple policy that enable the given callback
+    with the response.
+    """
+    def __init__(self, **kwargs):
+        self._cls = None
+    
+    def on_request(self, request):
+        # type: (PipelineRequest, Any) -> None
+        self._cls = request.context.options.pop('raw_response_hook', {})
```

Comment:
Why not looking at kwargs too? Should this be possible at the operation level?

Reply:
Where `raw_response_hook` is coming from?

Reply:
In the design, if users want to add custom hook, they need to enable custom_hook policy and then put their call back function in the kwargs with the key 'raw_response_hook'. In the pipeline, we interpret kwargs and store the information in request.context.options.

Reply:
The user should not have to provide a custom hook policy **and** provide the hook parameter.

Reply:
> Why not looking at kwargs too? Should this be possible at the operation level?

This should be possible on a per-invocation level.

Reply:
OK. We will update autorest to make it enabled by default.

Reply:
> > Why not looking at kwargs too? Should this be possible at the operation level?
> 
> This should be possible on a per-invocation level.

Yes. For every invocation, we will parse the kwargs and put the information in request.context.options

Reply:
In the initial design, I did (I insist on that since I don't know if you changed it), the **kwargs of "on_request" were the operation level of kwargs. It's why I was expecting to see:
```python
def on_request(self, request, **kwargs):
    self._cls = kwargs.pop('raw_response_hook', {})
```
I see in the ABC that "on_request" still has kwargs, are they same like before? Could both syntax work then?

Reply:
> ```python
> on_request
> ```

Yes, we changed the design a little bit. Now in pipeline.run, we parse the kwargs
def run(self, request, **kwargs):
        # type: (HTTPRequestType, Any) -> PipelineResponse
        context = PipelineContext(self._transport, **kwargs)
        pipeline_request = PipelineRequest(request, context)  # type: PipelineRequest[HTTPRequestType]
        first_node = self._impl_policies[0] if self._impl_policies else _TransportRunner(self._transport)
        return first_node.send(pipeline_request)  # type: ignore

and in policyrunner, we do not accept kwargs any longer
def send(self, request):
        # type: (PipelineRequest[HTTPRequestType], Any) -> PipelineResponse[HTTPRequestType, HTTPResponseType]
        self._policy.on_request(request)